server/routes/serviceProviderReferrals/serviceProviderReferralsController.ts

```diff
@@ -605,4 +609,76 @@ export default class ServiceProviderReferralsController {
 
     res.redirect(303, `/service-provider/end-of-service-report/${draftEndOfServiceReport.id}/outcomes/1`)
   }
+
+  async editEndOfServiceReportOutcome(req: Request, res: Response): Promise<void> {
+    const { accessToken } = res.locals.user.token
+
+    const endOfServiceReport = await this.interventionsService.getEndOfServiceReport(accessToken, req.params.id)
+    const referral = await this.interventionsService.getSentReferral(accessToken, endOfServiceReport.referralId)
+
+    const desiredOutcomeNumber = Number(req.params.number)
+
+    if (desiredOutcomeNumber > referral.referral.desiredOutcomesIds.length) {
+      throw createError(404, 'Outcome number is out of bounds')
+    }
+
+    const desiredOutcomeId = referral.referral.desiredOutcomesIds[desiredOutcomeNumber - 1]
+
+    const serviceCategory = await this.interventionsService.getServiceCategory(
+      accessToken,
+      referral.referral.serviceCategoryId
+    )
+
+    const desiredOutcome = serviceCategory.desiredOutcomes.find(val => val.id === desiredOutcomeId)
+
+    if (desiredOutcome === undefined) {
+      throw new Error(`Desired outcome for ID ${desiredOutcomeId} not found`)
+    }
+
+    const outcome = endOfServiceReport.outcomes.find(val => val.desiredOutcome.id === desiredOutcomeId) ?? null
+
+    let formValidationError: FormValidationError | null = null
+    let userInputData: Record<string, unknown> | null = null
+
+    if (req.method === 'POST') {
+      userInputData = req.body
+      const form = new EndOfServiceReportOutcomeForm(req, desiredOutcomeId, referral.referral.serviceUser)
+      const formData = await form.data()
+
+      if (formData.error) {
+        formValidationError = formData.error
+        res.status(400)
+      } else {
+        await this.interventionsService.updateDraftEndOfServiceReport(
+          accessToken,
+          endOfServiceReport.id,
+          formData.paramsForUpdate
+        )
+
+        const isLastDesiredOutcome = desiredOutcomeNumber === referral.referral.desiredOutcomesIds.length
```

Comment:
Not a big deal, but we could assign the `referral.referral.desiredOutcomesIds.length` to a `lastDesiredOutcome` variable and use it above when checking if the desired outcome is out of bounds.

Reply:
I'm not sure `lastDesiredOutcome` would make sense, since `.length` is a count — I've changed it to `numberOfDesiredOutcomesInReferral`. Unless I've missed what you meant?

Reply:
Sorry, I must've meant to say `lastDesiredOutcomeIndex` or something. I think what you've done works though!